### ISSUE TITLE
Improvements for mosaic detection

### ISSUE THREAD

**Issue by ladaapp** [ladaapp/lada#26]
Mosaic detection works pretty good already but there are a few things worth checking out

* Currently we use YOLO11m segmentation model as a base. All YOLO metrics are in the 99.9% range so let's try if we can get away with training a smaller model to improve runtime speed / lower compute requirements.
* recreate and retrain the model. while updating mosaic restoration model a new degradation option to encode clips twice to better work with low-quality/reencoded sources was introduced. assumption is that this would also increase detection accuracy of the mosaic detection model on such sources.

**Comment by ConnoisseurProtege** [comment]
I had some issues with mosaics like this(from https://github.com/ladaapp/lada/issues/36), where the mosaic is thinner/finer and it takes up a small portion of the image. Not sure if it's possible to improve the detection for this though as the program seems to downscale the image/vid first
![Image](https://github.com/user-attachments/assets/848c0843-cfb3-4499-97ea-c6b8c1b8fa76)

**Comment by ladaapp** [comment]
trained a new model based on YOLOv11n. After a lot more training and adjustments to the dataset I think this should be pretty much the same quality as the current bigger model.

It also will try to differentiate between nsfw and face mosaics and skips face mosaics in the restoration process. Doesn't work great yet but based on my testing it at least did not result in reducing the detection quality of nsfw mosaics so should be an improvement still.

@ConnoisseurProtege small mosaics should be reliably detected with the new and current detection model. I assume that in your scenario it was detected but the restoration model did not actually removed the mosaic pattern. If that is the case then it is a somewhat known issue currently, e.g. this ticket here https://github.com/ladaapp/lada/issues/40
In the GUI you can toggle between detection and restoration mode to verify if its a detection problem.

**Comment by ladaapp** [comment]
After a bit more testing with the new models:

* YOLOv11n struggles to detect bordering mosaic regions in some scenes compared to YOLOv11n-seg. So some pixels on the edge of the screen would not be restored. Otherwise are pretty much the same if it comes to misses. Segmentation variant is slightly slower.
* both new models both miss to detect mosaics in a few scenes where the current big model based on YOLOv11m-seg doesn't struggle. Not horrible though, it was noticeable on a specific video from my test set while I couldn't see a difference with all other videos tested.
* differentiation between nsfw and face mosaics can be useful to reduce artifacts and lower processing time when skipping those scenes but seems that training with both classes instead of a single mosaic class overall hurts the detection precision. I trained another small model with a single class and it could detect a few mosaics the other models trained on two classes missed.
* I noticed less false positives with the newer models compared to the old one. I assume this doesn't come from the model architecture but from the bigger and more diverse dataset the models have been trained on.
* speedup compared to current model is pretty huge

I will train another model based on v11s-seg with single class mosaics hoping this would achieve the same accuracy as the current model which is more or less perfect but pretty slow.
If that turns out correct I would ship with two-class v11n-seg and 1-class v11s-seg providing an accurate and a fast option. Default would probably the fast model as it works pretty well for most sources. 🤞

**Comment by ConnoisseurProtege** [comment]
@ladaapp Hey, I noticed that the wget command for the new detection model in the README doesn't actually download anything
Which parts do we need to edit also to get the new detection model to be used? In the gui it seems to still default to the old model, and I can't seem to change it. Do we need to specify it specifically in CLI?

**Comment by ladaapp** [comment]
oh, did not intend to commit the wget change yet. will revert back to previous link. thanks for notice.

Until then you have to adjust it manually after downloading it from the release:
* for cli pass the path to file via `--mosaic-detection-model-path`
* for gui replace filename `lada_mosaic_detection_model_v2.pt' manually to the new file ` lada_mosaic_detection_model_v3-yolo11n-beta.pt`

**Comment by seatv** [comment]
First of all, please tell me if I am posting in the wrong place.
I did a git clone last night and it pulled down the beta version. I was able to successfully run it on Windows.

First impression - on a eleven minute test clip I run, I found the resultant image quality to be very poor. Did not see any changes in mosaic detection, will look for one of the converted files with a recognition issue.

v2 file size for my test clip is 372MB, where as the v3 is 99.45MB. I did not specify any options in the command line:
```
lada>lada-cli --input D:\ToClean\IESM-066-ElevenMinuteClip.mp4 --output D:\Results\IESM-066-ElevenMinuteClip-Restored-v3.mp4 --device cuda:0
```

**Comment by ConnoisseurProtege** [comment]
> What do I need to specify in the command line to get the old video encoding settings back, as some changes were made to allow additional codecs?

My guess is that the crf set by default is the issue here. When I glanced through the changes it seems that the default from before was 20 and now it just goes with the encoder default( which is 23 default for libx264).  I also noticed that new detection model seems to produce a smaller output too with the same settings. Not sure why it's so much smaller in your case though

**Comment by ladaapp** [comment]
> First of all, please tell me if I am posting in the wrong place.

Yep, this ticket is about improvements for the mosaic detection model :-) 

Apparently PyAV does not set crf=23 for libx264 and equivalent for libx265 like ffmpeg but chooses some low-bitrate ABR setting... Fixed by setting conservative crf=20 for x264  and crf=23 for x265 as fallbacks if no --crf was provided.

> I also noticed that new detection model seems to produce a smaller output too with the same settings. Not sure why it's so much smaller in your case though

Hmm, not sure I understand what you mean by that. If this is also about encoder settings please open a new issue and provide output of `mediainfo` of the restored file.

**Comment by ConnoisseurProtege** [comment]
> Hmm, not sure I understand what you mean by that. If this is also about encoder settings please open a new issue and provide output of `mediainfo` of the restored file.

Ah I was just suggesting a reason why the output video might be smaller, the newer v11n-seg seems to miss some detections, resulting in a slightly smaller output(like a few MB smaller), but nothing as drastic as above.  

>Apparently PyAV does not set crf=23 for libx264 and equivalent for libx265 like ffmpeg but chooses some low-bitrate ABR setting... Fixed by setting conservative crf=20 for x264 and crf=23 for x265 as fallbacks if no --crf was provided.

Had no idea PyAV did this, thanks

**Comment by seatv** [comment]
Ok, getting back to the topic of this thread, I ran my eleven minute test clip with v3. For this particular clip, v2 never missed a single mosaic, found them all, even the one where the camera pan horizontally on a person on their back. However, v3 misses them. In the eleven minutes, there were several scenes where the mosaic was missed. [lada-cli on Tesla P-4 on an Ryzen 5 4650 GE processor].

Another thing I noticed is that the whole processing finishes in 20 minutes on this machine with v3 as opposed 27 minutes with v2.

**Comment by seatv** [comment]
My comments above was based on lada-cli invoked as follows:

```
D:\MyProjects\LadaEnv312\lada>lada-cli --input D:\ToClean\ElevenMinuteClip.mp4 --output D:\Results\ElevenMinuteClip-Restored-v3.mp4 --device cuda:0 --codec "h264" --crf "15" --preset "medium" --mosaic-detection-model-path "D:\MyProjects\LadaEnv312\lada\model_weights\lada_mosaic_detection_model_v3-yolo11n-beta.pt"
``` 

It would be nice to default to these, which were 0.6.0 defaults anyway. Those who want to override can override.  If not, I think a novice user will be left with a poor quality output.

**Comment by seatv** [comment]
@ladaapp, quick question - lada-cli --version says 0.6.0-dev. First I did the git clone, my assumption was that it would clone the latest, the --version gave me 0.6.0-dev, so I started over and this time got the zip file and that also gave me the same version number. Am I getting the Beta2? Should I be using the tag qualifier... then again zip file should give me the right version.

**Comment by seatv** [comment]
> noticed loss in accuracy of model released as beta1 was due to changes in dataset creation not the model architecture. Trained a new version on a newly created dataset which should hopefully bring it to an accuracy level as the currently used model has. (and it should be better on very-low quality content 480p content with encoding artifacts where the current model struggles with)
> 
> Released as beta2 https://github.com/ladaapp/lada/releases/tag/v0.7.0-beta2

This one does not miss any mosaics in my eleven minute test clip.

**Comment by ladaapp** [comment]
> This one does not miss any mosaics in my eleven minute test clip.

Thanks for testing.

Finished training of yet another version (beta3). Comparable to beta2 but less false-positive detections.
Doing some final smoke testing but this should be it. A lot faster compared to current model while also being more accurate and precise.
Filtering non-NSFW mosaics (faces/heads) works to some extent but does not reduce detection of NSFW mosaics so I'll keep it enabled as it can still be pretty useful on some sources even in its current state. I didn't expect too much given that it only present in 5% of my training data.

**Comment by ConnoisseurProtege** [comment]
> Finished training of yet another version (beta3). Comparable to beta2 but less false-positive detections.

I have an issue with the new model where it would detect the mosaic for a second or two, but then it will not detect it for a couple seconds, then back to detecting it in the same scene (ie camera angle is still the same, only slight movement, the mosaic never moves out of frame).

**Comment by JohnJimmy** [comment]
> > Finished training of yet another version (beta3). Comparable to beta2 but less false-positive detections.
> 
> I have an issue with the new model where it would detect the mosaic for a second or two, but then it will not detect it for a couple seconds, then back to detecting it in the same scene (ie camera angle is still the same, only slight movement, the mosaic never moves out of frame).

I get the same happening.  I thought it might be random so I ran the same video through lada twice and the mosaic showing and hiding happen in the exact same places, so I'm assuming it's to do with the model.

I don't want to sound like I'm complaining here, lada is far and above better than DeepMosaic - it's frankly incredible what lada can do with removing mosaics.  Thanks again for your work @ladaapp

**Comment by ladaapp** [comment]
released a slightly larger model v3.1-accurate as of lada v0.7.1 (still smaller than what we had in detection model v2) to address this regression -> https://github.com/ladaapp/lada/releases/tag/v0.7.1
